File: Courses/PHP Web Basics - май 2019/16.ExerciseEncapsulationInheritance/01.05.PizzaCalories/Dough.php
<?php


class Dough
{
    const CALORIES_PER_GRAM = 2;
    const DOUGH_TYPES = [
        "white" => 1.5,
	    "wholegrain" => 1.0
    ];
    const BAKING_TECHNIQUES = [
        "crispy" => .9,
        "chewy" => 1.1,
        "homemade" => 1.0
    ];
    /**
     * @var string
     */
    private $type;
    /**
     * @var string
     */
    private $bakingTechnique;
    /**
     * @var int
     */
    private $weight;

    /**
     * Dough constructor.
     * @param string $type
     * @param string $bakingTechnique
     * @param int $weight
     * @throws Exception
     */
    public function __construct(string $type, string $bakingTechnique, int $weight)
    {
        $this->setType($type);
        $this->setBakingTechnique($bakingTechnique);
        $this->setWeight($weight);
    }

    /**
     * @param string $type
     * @throws Exception
     */
    public function setType(string $type): void
    {
        if (!array_key_exists($type, self::DOUGH_TYPES)) {
            throw new \Exception("Invalid type of dough.");
        }
        $this->type = $type;
    }

    /**
     * @param string $bakingTechnique
     */
    public function setBakingTechnique(string $bakingTechnique): void
    {
        $this->bakingTechnique = $bakingTechnique;
    }

    /**
     * @param int $weight
     * @throws Exception
     */
    public function setWeight(int $weight): void
    {
        if ($weight < 1 || $weight > 200) {
            throw new \Exception("Dough weight should be in the range [1..200].");
        }
        $this->weight = $weight;
    }
    public function getCalories() : float {
    return $this->weight
        * self:: CALORIES_PER_GRAM
        * self:: DOUGH_TYPES[$this->type]
        * self:: BAKING_TECHNIQUES[$this->bakingTechnique];
    }
}

File: Courses/C# OOP/08. Encapsulation Exercise/08. Encapsulation Exercise/05.PizzaCalories/Dough.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace PizzaCalories
{
    public class Dough
    {
        private string flourType;
        private string bakingTechnique;
        private int weight;

        public Dough(string flourType, string bakingTechnique, int weight)
        {
            FlourType = flourType;
            BakingTechnique = bakingTechnique;
            this.Weight = weight;
        }

        public string FlourType
        {
            get => flourType; 
            set
            {
                if (value.ToLower() != "white" && value.ToLower() != "wholegrain")
                {
                    throw new ArgumentException("Invalid type of dough.");
                }

                flourType = value.ToLower();
            }
        }
     
        public string BakingTechnique
        {
            get => bakingTechnique; 
            set
            {

                if (value.ToLower() != "crispy" && value.ToLower() != "chewy" && value.ToLower() != "homemade")
                {
                    throw new ArgumentException("Invalid type of dough.");
                }
                bakingTechnique = value.ToLower();
            }
        }

        public int Weight
        {
            get => weight; 
            set
            {
                if (value < 1 || value > 200)
                {
                    throw new ArgumentException("Dough weight should be in the range [1..200].");
                }
                weight = value;
            }
        }

        public double DoughCalories { get => this.CalculateDoughCalories(); }

        public double CalculateDoughCalories()
        {
            double floorModifier = FlourType == "white" ? 1.5 : 1.00;
            double bakingModifier = BakingTechnique == "crispy" ? 0.9 : BakingTechnique == "chewy" ? 1.1: 1.0;
            return 2 * Weight * floorModifier * bakingModifier;
        }
    }

}
